clash-cores/src/Clash/Cores/Xilinx/DcFifo/BlackBoxes.hs

```diff
@@ -0,0 +1,215 @@
+{-|
+  Copyright   :  (C) 2022 Google Inc
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+  Blackbox implementation for primitives in "Clash.Cores.Xilinx.DcFifo.Explicit".
+-}
+
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Clash.Cores.Xilinx.DcFifo.BlackBoxes where
+
+import Prelude
+
+import Clash.Core.Literal (Literal(NaturalLiteral))
+import Clash.Core.TermLiteral (termToDataError, TermLiteralSNat(..))
+import Clash.Core.Term (Term(Literal))
+import qualified Clash.Primitives.DSL as DSL
+import Clash.Netlist.BlackBox.Types (BlackBoxFunction, emptyBlackBoxMeta)
+import Clash.Netlist.Types (TemplateFunction(..), BlackBox(BBFunction))
+import Clash.Netlist.Util (orNothing)
+
+import Clash.Netlist.Id (Identifier)
+import qualified Clash.Netlist.Id as Id
+import qualified Clash.Netlist.Types as N
+import qualified Clash.Netlist.BlackBox.Types as N
+
+import Control.Monad (unless)
+import Data.Either (lefts)
+import Data.Maybe (catMaybes)
+import GHC.Stack (HasCallStack)
+
+import Clash.Cores.Xilinx.DcFifo.Explicit (DcConfig(..))
+import Clash.Cores.Xilinx.DcFifo.Instances ()
+import Clash.Cores.Xilinx.Common (toTclBool, renderTcl, defIpConfig, IpConfig (properties))
+
+-- | Blackbox function for 'Clash.Cores.Xilinx.Fifo.dcFifo'. It parses the "DcConfig"
+-- supplied to 'dcFifo' from its Term representation, and passes them to two
+-- template functions:
+--
+--   * 'dcFifoTclTF': renders TCL file calling Xilinx's `create_ip`
+--   * 'dcFifoTF': instantiates IP generated in 'dcFifoTclTF'
+--
+-- Additionally, it generates an unique module name for the Xilinx IP.
+--
+dcFifoBBF :: HasCallStack => BlackBoxFunction
+dcFifoBBF _isD _primName args _resTys = do
+  let
+    [  knownNatN, _knownDomainWrite, _knownDomainRead, _knownNatDepth
+     , _constraint1, _constraint2
+     , either error id . termToDataError -> dcConfig
+     , _wClk, _rClk, _rst, _wData
+     , _rEnable
+
+    -- TODO: Make this blackbox return multiple results, instead of a tuple. See:
+    --       https://github.com/clash-lang/clash-compiler/pull/1560
+    --  , _, _, _, _, _, _, _
+     ] = lefts args
+
+  dcFifoName <- Id.makeBasic "dcfifo"
```

Comment:
Unfortunately we need a different definition of uniqueness than `makeBasic` gives us. Far from it: IP does not live in any library, it needs a globally unique name even in the face of multiple top entities. On the other hand, two identically configured IP instances are superfluous.

You don't need multiple top entities to get a collision with `makeBasic` though: `NOINLINE` is enough. So if I have two HDL files that both use a FIFO, both will get the name `dcfifo`, but those names collide. And on the other hand, if one HDL file has multiple identically configured FIFOs, it will generate an IP instance for each, whereas all of the generated code could share one IP instance that is instantiated multiple times in the HDL.

<details><summary>This example already generates name collisions</summary>

```haskell
module Collide where

import Clash.Prelude
import Clash.Cores.Xilinx.DcFifo

f ::
  Clock XilinxSystem ->
  Reset XilinxSystem ->
  Signal XilinxSystem (Maybe (BitVector 8)) ->
  Signal XilinxSystem Bool ->
  Signal XilinxSystem (Maybe (BitVector 8)) ->
  Signal XilinxSystem Bool ->
  ( XilinxFifo XilinxSystem XilinxSystem 5 8
  , XilinxFifo XilinxSystem XilinxSystem 10 8)
f clk rst wg rEnG wh rEnH = (outG, outH)
 where
  outG = g clk clk rst wg rEnG
  outH = h clk clk rst wh rEnH
{-# NOINLINE f #-}
{-# ANN f (defSyn "f") #-}

g ::
  Clock XilinxSystem ->
  Clock XilinxSystem ->
  Reset XilinxSystem ->
  Signal XilinxSystem (Maybe (BitVector 8)) ->
  Signal XilinxSystem Bool ->
  XilinxFifo XilinxSystem XilinxSystem 5 8
g = dcFifo defConfig
{-# NOINLINE g #-}
-- {-# ANN g (defSyn "g") #-}

h ::
  Clock XilinxSystem ->
  Clock XilinxSystem ->
  Reset XilinxSystem ->
  Signal XilinxSystem (Maybe (BitVector 8)) ->
  Signal XilinxSystem Bool ->
  XilinxFifo XilinxSystem XilinxSystem 10 8
h = dcFifo defConfig
{-# NOINLINE h #-}
-- {-# ANN h (defSyn "h") #-}
```
</details>

While it generates ugly names, I'd prefer this to be based on the `~INCLUDENAME`, because we deduplicate those.<sup>†</sup> And for the long term, we should probably add a mechanism to do this without generating names with hashes.

<sup>†</sup> Frankly Martijn and I are not that clear on how these deduplication unique hashes work, and they are worrying, but that's outside the scope of this PR.